src/internal/pfsdb/branches.go

```diff
@@ -528,28 +535,92 @@ func DeleteBranchTrigger(ctx context.Context, tx *pachsql.Tx, from BranchID) err
 	return nil
 }
 
-func fetchBranchInfoByBranch(ctx context.Context, tx *pachsql.Tx, branch *Branch) (*pfs.BranchInfo, error) {
+func fetchBranchInfoByBranch(ctx context.Context, ext sqlx.ExtContext, branch *Branch) (*pfs.BranchInfo, error) {
 	if branch == nil {
 		return nil, errors.Errorf("branch cannot be nil")
 	}
 	branchInfo := &pfs.BranchInfo{Branch: branch.Pb(), Head: branch.Head.Pb()}
 	var err error
-	branchInfo.DirectProvenance, err = GetDirectBranchProvenance(ctx, tx, branch.ID)
+	branchInfo.DirectProvenance, err = GetDirectBranchProvenance(ctx, ext, branch.ID)
 	if err != nil {
 		return nil, errors.Wrap(err, "could not get direct branch provenance")
 	}
-	branchInfo.Provenance, err = GetBranchProvenance(ctx, tx, branch.ID)
+	branchInfo.Provenance, err = GetBranchProvenance(ctx, ext, branch.ID)
 	if err != nil {
 		return nil, errors.Wrap(err, "could not get full branch provenance")
 	}
-	branchInfo.Subvenance, err = GetBranchSubvenance(ctx, tx, branch.ID)
+	branchInfo.Subvenance, err = GetBranchSubvenance(ctx, ext, branch.ID)
 	if err != nil {
 		return nil, errors.Wrap(err, "could not get full branch subvenance")
 	}
 	// trigger info
-	branchInfo.Trigger, err = GetBranchTrigger(ctx, tx, branch.ID)
+	branchInfo.Trigger, err = GetBranchTrigger(ctx, ext, branch.ID)
 	if err != nil {
 		return nil, errors.Wrap(err, "could not get branch trigger")
 	}
 	return branchInfo, nil
 }
+
+// Helper functions for watching branches.
+type branchUpsertHandler func(id BranchID, branchInfo *pfs.BranchInfo) error
+type branchDeleteHandler func(id BranchID) error
+
+func WatchBranchesInRepo(ctx context.Context, db *pachsql.DB, listener collection.PostgresListener, repoID RepoID, onUpsert branchUpsertHandler, onDelete branchDeleteHandler) error {
+	watcher, err := postgres.NewWatcher(db, listener, randutil.UniqueString(fmt.Sprintf("watch-branches-in-repo-%d", repoID)), BranchesInRepoChannel(repoID))
+	if err != nil {
+		return err
+	}
+	defer watcher.Close()
+	// Optimized query for getting branches in a repo.
+	query := getBranchBaseQuery + fmt.Sprintf("\nWHERE %s = ?\nORDER BY %s ASC", BranchColumnRepoID, BranchColumnID)
+	query = db.Rebind(query)
```

Comment:
why do you need to rebind the query here?

Reply:
to convert the `?` into postgres `$x` style.